### ISSUE TITLE
[Feature Request] New option `--assets-base`

### ISSUE THREAD

**Issue by brillout** [evanw/esbuild#2707]
The new option `--assets-base` would behave like [`--public-path`](https://esbuild.github.io/api/#public-path) while skipping JavaScript imports.

Let me know in case you wonder what the use case for this is. (I'm building a framework on top of Vite + vite-plugin-ssr.)

Thank you for building esbuild, it's an amazing tool.

**Comment by evanw** [comment]
Can you describe your problem instead of your proposed solution? What problem are you having? What are you trying to do?

**Comment by brillout** [comment]
Imagine a library using esbuild to build its source codes and that contains both JavaScript and SVG images.

When Vite sees `import "/assets/chunk-TTLAZ2T2.js"` then Vite won't be able to resolve `/assets/chunk-TTLAZ2T2.js`. (Same for Node.js btw.)

Vite can only handle the original relative path `import "../chunk-TTLAZ2T2.js"`.

Now Vite serves assets under `/assets/` and that's why asset references need to have the `/assets/` prefix.

In a nutshell: JavaScript import paths need to be relative, while the path assets references need to be prefixed with `/assets/`.

Let me know if my example isn't clear, I'm happy to elaborate. (My concrete situation is slightly different, but this example is simpler to convey.)

**Comment by evanw** [comment]
This sounds like a problem with Vite to me? If they decided to have things work this way then I would expect for them to have figured out some way for this to work. How are other Vite projects not broken due to this problem as well?

**Comment by brillout** [comment]
Simpler example: https://github.com/brillout/esbuild-reprod-2707. Let me know if you want me to elaborate.

> How are other Vite projects not broken due to this problem as well?

I guess because Vite doesn't use `publicPath`.

**Comment by nitedani** [comment]
Why is the public path prepended to a dynamic import, even when I set platform:"node"? Why does esbuild assume that the code will be run in the browser?

**Comment by brillout** [comment]
> Why is the public path prepended to a dynamic import, even when I set platform:"node"? Why does esbuild assume that the code will be run in the browser?

Note that changing the behavior of `platform: "node"` doesn't suffice to fix the problem. What should esbuild do when `platform` is `"neutral"`? Even with `platform: "browser"` there are situations when you don't want esbuild to prepend `/assets/` to JavaScript import paths.

Imagine a library on npm that was bundled using esbuild.  Now imagine a user app using another bundler (Vite, Rollup, ...) which uses that library. The user's bundler will most likely expect all JavaScript import paths to be relative and any absolute JavaScript import path will choke the user's bundler.

That's why I propose a new option. AFAICT it's the only way to fix the situation: what should happen really depends on the use case.

**Comment by brillout** [comment]
Actually, thinking of it, I think a new option `--preserve-relative-paths` (`boolean`) would be cleaner than `--assets-base`.

**Comment by brillout** [comment]
Also for assets referenced in CSS, actually. I'd suggest to extend `--preserve-relative-paths` to be a `string`, for example `--preserve-relative-paths css,js`:

```css
/* node_modules/some-ui-library/style.css */

button::before {
  /* The URL wasn't replaced with '/assets/some-icon.svg'
     because we used `--preserve-relative-paths css` */
  background: url('./some-icon.svg')
}
```

Now imagine a Vite user who is using `some-ui-library`. Vite can only process `url('./some-icon.svg')`, hence the need for `--preserve-relative-paths css`.

**Comment by brillout** [comment]
Closing in favor of https://github.com/evanw/esbuild/issues/2981.

Apologies for the convoluted explanation here. I hope the new ticket makes a better job at explaining the problem.
